### PR TITLE
Immutable HTTP APIs, part 1

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3855

I moved some existing tests into `integration/`.

In general, I don't see much point in unit testing the server/client separate from each other, except in edge cases that are hard to test with a real client.

### CODE REVIEW

File: integration/test_storage_http.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        expected_version = self.http.storage_server.get_version()
+        expected_version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"available-space"
+        )
+        expected_version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"maximum-immutable-share-size"
+        )
+        self.assertEqual(version, expected_version)
+
+
```
Comment (itamarst) [inline integration/test_storage_http.py:156]: Everything before this point in the module was moved from `src/allmydata/test/test_storage_http.py`, pretty much unchanged.

Comment (itamarst) [reply integration/test_storage_http.py:156]: Everything after this point is new code.

File: docs/proposed/http-storage-node-protocol.rst [hunk 2]
```diff
@@ -382,6 +382,11 @@ the server will respond with ``400 BAD REQUEST``.
 
 If authorization using the secret fails, then a ``401 UNAUTHORIZED`` response should be sent.
 
+Encoding
+~~~~~~~~
+
+* ``storage_index`` should be base32 encoded (RFC3548) in URLs.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:388]: It was motivated by their existence, yes. But easy enough to use something else, I don't have a strong opinion, so would be good to know which you prefer.

File: integration/test_storage_http.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            downloaded = yield im_client.read_share_chunk(
+                storage_index, 1, offset, length
+            )
+            self.assertEqual(downloaded, expected_data[offset : offset + length])
+
+    def test_multiple_shares_uploaded_to_different_place(self):
+        """
+        If a storage index has multiple shares, uploads to different shares are
+        stored separately and can be downloaded separately.
+        """
```
[the remark this answers is not in the document]
Comment (itamarst) [reply integration/test_storage_http.py:234]: Fill them in in next PR.

File: src/allmydata/storage/http_server.py [hunk 4]
```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
...
+    @_authorized_route(
+        _app,
+        {Secrets.LEASE_RENEW, Secrets.LEASE_CANCEL, Secrets.UPLOAD},
+        "/v1/immutable/<string:storage_index>",
+        methods=["POST"],
+    )
+    def allocate_buckets(self, request, authorization, storage_index):
+        """Allocate buckets."""
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        info = loads(request.content.read())
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:179]: Some future PR. I can do it next if you want to see it integrated sooner.

File: src/allmydata/storage/http_server.py [hunk 5]
```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
...
+        methods=["PATCH"],
+    )
+    def write_share_data(self, request, authorization, storage_index, share_number):
+        """Write data to an in-progress immutable upload."""
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        content_range = request.getHeader("content-range")
+        if content_range is None:
+            offset = 0
+        else:
+            offset = int(content_range.split()[1].split("-")[0])
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:227]: Looks like werkzeug has one.

File: src/allmydata/storage/http_server.py [hunk 6]
```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
...
+        # TODO if raises ConflictingWriteError, return HTTP CONFLICT code.
+
+        if finished:
+            bucket.close()
+            request.setResponseCode(http.CREATED)
+        else:
+            request.setResponseCode(http.OK)
+
+        # TODO spec says we should return missing ranges. but client doesn't
+        # actually use them? So is it actually useful?
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:250]: Ah, yes.

File: src/allmydata/storage/immutable.py [hunk 7]
```diff
@@ -375,13 +375,16 @@ def __init__(self, ss, incominghome, finalhome, max_size, lease_info, clock):
...
+        """
+        Write data at given offset, return whether the upload is complete.
+        """
         # Delay the timeout, since we received data:
         self._timeout.reset(30 * 60)
         start = self._clock.seconds()
         precondition(not self.closed)
         if self.throw_out_all_data:
-            return
+            return False
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/immutable.py:387]: I'll file a follow-up ticket.

File: src/allmydata/storage/http_client.py [hunk 8]
```diff
@@ -86,7 +125,146 @@ def get_version(self):
...
+        response = yield self._client._request(
+            "PATCH",
+            url,
+            upload_secret=upload_secret,
+            data=data,
+            headers=Headers(
+                {
+                    # The range is inclusive, thus the '- 1'. '*' means "length
+                    # unknown", which isn't technically true but adding it just
+                    # makes things slightly harder for calling API.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:214]: I'll expand the comment.

**Comment by itamarst** [comment]
I do not think plan with testing will work. The test code cannot run synchronously, basically.

File: integration/test_storage_http.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply integration/test_storage_http.py:234]: I opened ticket and added comments referring to it.

**Comment by itamarst** [comment]
Latest push is still work-in-progress with many review comments to address.

File: src/allmydata/storage/http_client.py [hunk 9]
```diff
@@ -86,7 +125,146 @@ def get_version(self):
...
+            {"share-numbers": share_numbers, "allocated-size": allocated_size}
+        )
+        response = yield self._client._request(
+            "POST",
+            url,
+            lease_renew_secret=lease_renew_secret,
+            lease_cancel_secret=lease_cancel_secret,
+            upload_secret=upload_secret,
+            data=message,
+            headers=Headers({"content-type": ["application/cbor"]}),
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:176]: Filed.

File: src/allmydata/storage/http_client.py [hunk 10]
```diff
@@ -59,25 +78,45 @@ def __init__(
...
         return headers
 
-    def _request(self, method, url, secrets, **kwargs):
+    def _request(
+        self,
+        method,
+        url,
+        lease_renew_secret=None,
+        lease_cancel_secret=None,
+        upload_secret=None,
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:101]: Yeah it's a private method, can always update it later.

File: src/allmydata/storage/http_client.py [hunk 11]
```diff
@@ -86,7 +125,146 @@ def get_version(self):
...
+    @inlineCallbacks
+    def create(
+        self,
+        storage_index,
+        share_numbers,
+        allocated_size,
+        upload_secret,
+        lease_renew_secret,
+        lease_cancel_secret,
+    ):  # type: (bytes, List[int], int, bytes, bytes, bytes) -> Deferred[ImmutableCreateResult]
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:151]: Done.

File: src/allmydata/storage/http_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:250]: After further thought:

1. The client needs to keep track of what it's sent anyway, just to make sure it's all uploaded (it shouldn't rely on this, because that introduces a bunch more latency).
2. If the client crashes, it needs to track some persistent state anyway (upload secret).
3. The failure mode for not having this is... uploading a bit extra. But the code gets simpler on both client and server.

So I propose just not doing this, and if it ever turns out to be, useful we can add an extra endpoint which is "show status of in-progress upload".

File: src/allmydata/storage/http_server.py [hunk 12]
```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
...
+        """Write data to an in-progress immutable upload."""
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        content_range = request.getHeader("content-range")
+        if content_range is None:
+            offset = 0
+        else:
+            offset = int(content_range.split()[1].split("-")[0])
+
+        # TODO basic checks on validity of start, offset, and content-range in general. also of share_number.
+        # TODO basic check that body isn't infinite. require content-length? or maybe we should require content-range (it's optional now)? if so, needs to be rflected in protocol spec.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:230]: If you're uploading a specific length, you don't need chunked encoding? So I think it's ok to require content-length.

**Comment by itamarst** [comment]
This is ready for another round of review.

* Most comments have been addressed, including testing APIs.
* Others require further discussion; see unresolved ones above.

File: src/allmydata/storage/http_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/storage/http_server.py:253]: (1) makes sense - albeit, for in-memory tracking.  The client probably wants to start uploading the 2nd chunk before the 1st finishes (maybe it even keeps N of them running in parallel all the time).  However, I don't think that (2) follows.  I'm still pretty sure that the upload secret is going to be a deterministic function of other local state the client holds.  Do we just need to design that to get rid of the uncertainty about it?  So, given this information in the response, I don't believe the client will actually need persistent state to be able to resume an upload.  Beyond that, even if the client keeps local state, it has no way of knowing whether the server has decided to give up on the transfer and throw away the data or not.  If it uses only local persistent state then it will "finish" the upload while the beginning is now missing.  It can probably infer that this is the case when the last write doesn't return a CREATED code but that's extra client side logic too.

For (3) we can't say if it is a little bit or not.  If I have a 20GB share to upload and I get halfway through, it would be pretty sad for my client to have to start over again when we could pretty easily avoid it.  To take this to the extreme, maybe my setup is flaky enough that I can *never* complete an upload in one try.  Persistent local state is another solution to this but between "the function returns enough information to finish the process" and "keep local state" my personal preference is definitely for the former because it introduces fewer difficult requirements on the client.

Does the server get that much more complicated by having to return this information?  Doesn't it already have to keep track of it to make sure that no single location in the sharefile can have different bytes written to it?

File: docs/proposed/http-storage-node-protocol.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (exarkun) [reply docs/proposed/http-storage-node-protocol.rst:388]: I'm not sure.  Absent other considerations, base64url seems better because it's more compact and because we've started adopting it elsewhere.

On the other hand, consistency with other encoding of storage-index in the codebase doesn't seem like a bad thing.  Storage indexes aren't the *most* user-facing thing but they do show up here and there ... and for an admin you're probably going to run into them.  Having to deal with two different formats seems like it would be a drag.

File: src/allmydata/storage/http_server.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/storage/http_server.py:233]: Ah yea, good point.

**Review by exarkun — APPROVED** [review]
Thanks.  Looks good.  Some minor comments inline.  The only one that might be actionable is the one about returning that extra info from `write_share_data` - but maybe that's follow-up work anyway.

File: docs/proposed/http-storage-node-protocol.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (meejah) [reply docs/proposed/http-storage-node-protocol.rst:388]: Consistency would be good here when debugging any differences between this and Foolscap (right?).

[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:388]: OK, leaving unchanged.

File: src/allmydata/storage/http_server.py [hunk 13]
```diff
@@ -154,3 +167,121 @@ def _cbor(self, request, data):
...
+        # TODO in https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3860
+        # 1. basic checks on validity on storage index, share number
+        # 2. missing range header should have response code 200 and return whole thing
+        # 3. malformed range header should result in error? or return everything?
+        # 4. non-bytes range results in error
+        # 5. ranges make sense semantically (positive, etc.)
+        # 6. multiple ranges fails with error
+        # 7. missing end of range means "to the end of share"
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        range_header = parse_range_header(request.getHeader("range"))
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:273]: Yeah, just need to additionally check for "there is no header" case ourselves and that should do it.

File: src/allmydata/storage/http_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:253]: It's not that big a deal, I'll add it. The only potential problem is https://github.com/mlenzen/collections-extended/issues/172 and looks like that'll be fixed in near future.

File: src/allmydata/storage/http_server.py [hunk 14]
```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
...
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        info = loads(request.content.read())
+        upload_key = authorization[Secrets.UPLOAD]
+
+        if storage_index in self._uploads:
+            # Pre-existing upload.
+            in_progress = self._uploads[storage_index]
+            if in_progress.upload_key == upload_key:
+                # Same session.
+                # TODO add BucketWriters only for new shares
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_server.py:190]: Going to do that next, but I expanded a bit at least.

**Comment by itamarst** [comment]
Still need to figure out:

* [ ] Why RangeMap is failing in bogus way (might need to file another bug).
* [ ] Get another PR merged removing Python 3.6 support (so we can get newer versions of collection_extended).

**Comment by paulwratt** [comment]
ready to go by the looks of it (CI)
